lib/util/cmcd_batching_manager.js

```diff
@@ -0,0 +1,296 @@
+      timer.stop();
+    }
+    for (const timer of this.retryTimers_.values()) {
+      timer.stop();
+    }
+    this.batches_.clear();
+    this.timers_.clear();
+    this.retryQueue_ = [];
+    this.retryTimers_.clear();
+    this.goneUrls_.clear();
+  }
+
+  /**
+   * Generate a unique key for a target configuration.
+   *
+   * @param {!shaka.extern.CmcdTarget} target
+   * @return {string}
+   * @private
+   */
+  generateTargetKey_(target) {
+    return JSON.stringify({
+      url: target.url,
+      mode: target.mode,
+      batchSize: target.batchSize,
+      batchTimer: target.batchTimer,
+    });
+  }
+
+  /**
+   * Flush a batch by its target key.
+   *
+   * @param {string} key
+   * @private
+   */
+  flushByTargetKey_(key) {
+    const batch = this.batches_.get(key);
+    if (!batch || batch.cmcdData.length === 0) {
+      return;
+    }
+
+    const {target, cmcdData} = batch;
+    const retryParams = shaka.net.NetworkingEngine.defaultRetryParameters();
+
+    const request = shaka.net.NetworkingEngine.makeRequest(
+        [target.url], retryParams);
+    request.method = 'POST';
+    request.body = shaka.util.StringUtils.toUTF8(cmcdData);
+    request.headers['Content-Type'] = 'txt/cmcd';
+
+    this.sendBatchReport_(request)
+        .then((response) => {
+          if (response && response.status === 410) {
+            this.goneUrls_.add(target.url);
+            shaka.log.info('CMCD endpoint returned 410 Gone, ' +
+                'removing from future batches:', target.url);
+          } else if (response && response.status === 429) {
+            // Rate limited, add to retry queue
+            this.retryQueue_.push({
+              request: request,
+              retryCount: 0,
+              sendTime: Date.now() + this.retryDelays_[0],
+            });
+            if (this.retryQueue_.length === 1) {
+              const timer = new shaka.util.Timer(() => {
+                this.processRetryQueue_();
+              });
+              timer.tickAfter(this.retryDelays_[0] / 1000);
+              this.retryTimers_.set('retry', timer);
+            }
+          }
+        })
+        .catch((error) => {
+          shaka.log.warning('Failed to send CMCD batch report:', error);
+        });
+
+    // Clear the batch data
+    batch.cmcdData = '';
+  }
+
+  /**
+   * Send a batch report request.
+   *
+   * @param {!shaka.extern.Request} request
+   * @return {!Promise<!shaka.extern.Response>}
+   * @private
+   */
+  async sendBatchReport_(request) {
+    const requestType = shaka.net.NetworkingEngine.RequestType.CMCD;
+    const networkingEngine = this.player_.getNetworkingEngine();
+
+    try {
+      const operation = networkingEngine.request(requestType, request);
+      const response = await operation.promise;
+      return response;
+    } catch (error) {
+      // Return the response from the error if available
+      if (error.response) {
+        return error.response;
```

Comment:
I don't think we ever attach Response object to error directly, see: https://github.com/shaka-project/shaka-player/blob/main/lib/net/http_plugin_utils.js#L59-L68